File: 155.min-stack.js
/*
 * @lc app=leetcode id=155 lang=javascript
 *
 * [155] Min Stack
 *
 * https://leetcode.com/problems/min-stack/description/
 *
 * algorithms
 * Easy (39.00%)
 * Likes:    2172
 * Dislikes: 237
 * Total Accepted:    357.4K
 * Total Submissions: 913.7K
 * Testcase Example:  '["MinStack","push","push","push","getMin","pop","top","getMin"]\n[[],[-2],[0],[-3],[],[],[],[]]'
 *
 * Design a stack that supports push, pop, top, and retrieving the minimum
 * element in constant time.
 * 
 * 
 * push(x) -- Push element x onto stack.
 * pop() -- Removes the element on top of the stack.
 * top() -- Get the top element.
 * getMin() -- Retrieve the minimum element in the stack.
 * 
 * 
 * 
 * 
 * Example:
 * 
 * 
 * MinStack minStack = new MinStack();
 * minStack.push(-2);
 * minStack.push(0);
 * minStack.push(-3);
 * minStack.getMin();   --> Returns -3.
 * minStack.pop();
 * minStack.top();      --> Returns 0.
 * minStack.getMin();   --> Returns -2.
 * 
 * 
 * 
 * 
 */

// @lc code=start
/**
 * initialize your data structure here.
 */
var MinStack = function() {
     this.content = [];
     this.mins = [];
};

/** 
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function(x) {
    this.content.push(x);
    if(!this.mins.length){
        this.mins.push(x);
    }else{
        if(this.mins[this.mins.length - 1] > x){
            this.mins.push(x);
        }else{
            this.mins.push(this.mins[this.mins.length - 1]);
        }
    }
    

    
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    this.content.pop();
    this.mins.pop();
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    if(this.content.length)
        return this.content[this.content.length - 1];
    else
        return null;
    
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    return this.mins[this.mins.length - 1];
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */
// @lc code=end

File: java/155.min-stack.java
/*
 * @lc app=leetcode id=155 lang=java
 *
 * [155] Min Stack
 */

// @lc code=start
class Cell{
    int val;
    int min;
    Cell(int v, int min){
        this.val = v;
        this.min = min;

    }
}
class MinStack {

    /** initialize your data structure here. */
    List<Cell> data;
    public MinStack() {
         data = new LinkedList<Cell>();
    }
    
    public void push(int x) {
        int min =x;
        if(data.size()!= 0){
            min = Math.min(min, data.get(data.size()-1).min);
        }
        data.add(new Cell(x, min));
        
    }
    
    public void pop() {
        data.remove(data.size()-1);
    }
    
    public int top() {
        return data.get(data.size()-1).val;
        
    }
    
    public int getMin() {
        return  data.get(data.size()-1).min;
        
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */
// @lc code=end

